File: lib/javascript/src/expression/ArrowFunctionExpression.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "javascript/inc/javascript.h"
#include "javascript/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "javascript/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace javascript { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace expression { 
  ArrowFunctionExpression::ArrowFunctionExpression(NodeId _id, Factory *_factory) :
         Positioned(_id, _factory),
    Expression(_id, _factory),
    Function(_id, _factory),
    m_expression(false)
  {
  }

  ArrowFunctionExpression::~ArrowFunctionExpression() {
  }

  void ArrowFunctionExpression::prepareDelete(bool tryOnVirtualParent){
    if (tryOnVirtualParent) {
      base::Positioned::prepareDelete(false);
    }
    expression::Expression::prepareDelete(false);
    statement::Function::prepareDelete(false);
  }

  NodeKind ArrowFunctionExpression::getNodeKind() const {
    return ndkArrowFunctionExpression;
  }

  bool ArrowFunctionExpression::getExpression() const {
    return m_expression;
  }

  void ArrowFunctionExpression::setExpression(bool _expression) {
    m_expression = _expression;
  }

  bool ArrowFunctionExpression::setEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    if (tryOnVirtualParent) {
      if (base::Positioned::setEdge(edgeKind, edgeEnd, false)) {
        return true;
      }
    }
    if (expression::Expression::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    if (statement::Function::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  bool ArrowFunctionExpression::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    if (tryOnVirtualParent) {
      if (base::Positioned::removeEdge(edgeKind, edgeEnd, false)) {
        return true;
      }
    }
    if (expression::Expression::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    if (statement::Function::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void ArrowFunctionExpression::accept(Visitor &visitor) const {
    visitor.visit(*this);
  }

  void ArrowFunctionExpression::acceptEnd(Visitor &visitor) const {
    visitor.visitEnd(*this);
  }

  double ArrowFunctionExpression::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      const ArrowFunctionExpression& node = dynamic_cast<const ArrowFunctionExpression&>(base);
      double matchAttrs = 0;
      if(node.getGenerator() == getGenerator()) ++matchAttrs;
      if(node.getAsync() == getAsync()) ++matchAttrs;
      if(node.getExpression() == getExpression()) ++matchAttrs;
      return matchAttrs / (3 / (1 - Common::SimilarityMinimum)) + Common::SimilarityMinimum;
    } else {
      return 0.0;
    }
  }

  void ArrowFunctionExpression::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

  }

  NodeHashType ArrowFunctionExpression::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "expression::ArrowFunctionExpression", strlen("expression::ArrowFunctionExpression"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void ArrowFunctionExpression::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    if (withVirtualBase)
      Positioned::save(binIo,false);

    Expression::save(binIo,false);

    Function::save(binIo,false);

    unsigned char boolValues = 0;
    boolValues <<= 1;
    if (m_expression) 
      boolValues |= 1;
    binIo.writeUByte1(boolValues);

  }

  void ArrowFunctionExpression::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    if (withVirtualBase)
      Positioned::load(binIo, false);

    Expression::load(binIo,false);

    Function::load(binIo,false);

    unsigned char boolValues = binIo.readUByte1();
    m_expression = boolValues & 1;
    boolValues >>= 1;

  }


}


}}}

File: javascript/cl/JSAN/src/ast/expression/arrowFunctionExpression.js
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

import * as globals from '../../globals.js';

const factory = globals.getFactory();

export default function (node, parent, firstVisit) {
    if (firstVisit) {
        if (globals.getWrapperOfNode(node) !== undefined) {
            return;
        }
        const arrowFunctionExpression = factory.createArrowFunctionExpressionWrapper();
        globals.setPositionInfo(node, arrowFunctionExpression);
        arrowFunctionExpression.setExpression(node.expression);
        arrowFunctionExpression.setGenerator(node.generator);
        arrowFunctionExpression.setAsync(node.async);
        return arrowFunctionExpression;
    } else {
        const arrowFunctionExpressionWrapper = globals.getWrapperOfNode(node);

        if (node.params != null) {
            for (let i = 0; i < node.params.length; i++) {
                if (node.params[i] != null) {
                    globals.safeSet(arrowFunctionExpressionWrapper, "addParams", node.params[i], "ARROWFUNCTIONEXPRESSION - Could not add param!");
                }
            }
        }

        if (node.body != null) {
            globals.safeSet(arrowFunctionExpressionWrapper, "setBody", node.body, "ARROWFUNCTIONEXPRESSION - Could not set body!");
        }

        if (node.id != null) {
            globals.safeSet(arrowFunctionExpressionWrapper, "setIdentifier", node.id, "ARROWFUNCTIONEXPRESSION - Could not set identifier!");
        }
    }
}
